### PR TITLE
Extend PreModuleImportCallback to support bridging header compilation

### PR DESCRIPTION
**Pull request by adrian-prantl**

### CODE REVIEW

**Comment by adrian-prantl** [comment]
test with https://github.com/apple/llvm-project/pull/8103
@swift-ci test

**Review by chelcassanova — APPROVED** [review]
This makes the module import callback a lot cleaner, LGTM!

File: lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -761,10 +761,15 @@ ASTContext::~ASTContext() {
...
 void ASTContext::SetPreModuleImportCallback(
-    std::function<void(llvm::StringRef ModuleName, bool IsOverlay)> callback) {
+    PreModuleImportCallbackPtr callback) {
   PreModuleImportCallback = callback;
 }
 
+ASTContext::PreModuleImportCallbackPtr
+ASTContext::GetPreModuleImportCallback() const {
+  return PreModuleImportCallback;
+}
```
Comment (kastiglione) [inline lib/AST/ASTContext.cpp:771]: What do you think of this? To move responsibility of the null check (of the callback) out of the caller's hands.

```suggestion
void ASTContext::PreModuleImport(llvm::StringRef ModuleName, ModuleImportKind kind) const {
  if (PreModuleImportCallback)
    PreModuleImportCallback(ModuleName, kind);
}
```

File: include/swift/AST/ASTContext.h [hunk 2]
```diff
@@ -295,10 +295,19 @@ class ASTContext final {
   /// OutputBackend for writing outputs.
   llvm::IntrusiveRefCntPtr<llvm::vfs::OutputBackend> OutputBackend;
 
+  enum ModuleImportKind {
+    Module = 0,
+    Overlay,
+    BridgingHeader
+  };
+  using PreModuleImportCallbackPtr =
+      std::function<void(llvm::StringRef ModuleName, ModuleImportKind kind)>;
```
Comment (kastiglione) [inline include/swift/AST/ASTContext.h:304]: one of these variables is capitalized, the other isn't.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci smoke test

File: lib/AST/ASTContext.cpp [hunk 3]
```diff
@@ -2476,7 +2482,7 @@ ASTContext::getModule(ImportPath::Module ModulePath, bool AllowMemoryCached) {
 
   auto moduleID = ModulePath[0];
   if (PreModuleImportCallback)
-    PreModuleImportCallback(moduleID.Item.str(), false /*=IsOverlay*/);
+    PreModuleImportCallback(moduleID.Item.str(), ModuleImportKind::Module);
```
Comment (kastiglione) [inline lib/AST/ASTContext.cpp:2485]: You could remove the `if (PreModuleImportCallback)` and call the Hook:

```c++
PreModuleImportHook(moduleID.Item.str(), ModuleImportKind::Module);
```

**Review by kastiglione — APPROVED** [review]

**Comment by adrian-prantl** [comment]
@swift-ci smoke test

**Comment by medismailben** [comment]
Great! LGTM!

**Review by medismailben — APPROVED** [review]

**Pull request merged**